File: kontrola_tymu.py
import sys
import requests
from bs4 import BeautifulSoup
import urllib.parse

# 1. Kontrola a načtení PINu z parametrů při spuštění
if len(sys.argv) < 2:
    print("Chyba: Nebyl zadán PIN.")
    sys.exit(1)

pin = sys.argv[1]

# Společné hlavičky pro web
headers_web = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
}

# --- ČÁST A: Získání týmů z webu (upcoming_teams) ---
print("Stahuji seznam týmů z webu (rezervace)...")
url_web = "https://www.hospodskykviz.cz/hospody/u-posty-vinohrady-st/rezervace"

upcoming_teams = []
response_web = requests.get(url_web, headers=headers_web)

if response_web.status_code == 200:
    soup = BeautifulSoup(response_web.text, 'html.parser')
    tables = soup.find_all('table')
    
    for table in tables:
        first_row = table.find('tr')
        if first_row and 'Název týmu' in first_row.text:
            rows = table.find_all('tr')[1:]
            for row in rows:
                cols = row.find_all(['td', 'th'])
                if len(cols) >= 2:
                    team_name = cols[0].text.strip()
                    if team_name != "Toto místo je zatím volné.":
                        upcoming_teams.append(team_name)
            break
else:
    print(f"Chyba při stahování webu. Kód: {response_web.status_code}")
    sys.exit(1)


# --- ČÁST B: Získání týmů z API pomocí PINu (registered_teams) ---
print(f"Stahuji seznam týmů z API pro PIN: {pin} (simuluji mobilní Chrome)...")
url_api = "https://www.hospodskykviz.cz/api/loadquiz"

headers_api = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Origin": "https://hodnoceni.hospodskykviz.cz",
    "Accept": "application/json, text/javascript, */*; q=0.01"
}
payload = {
    "pin": pin
}

registered_teams = []
response_api = requests.post(url_api, headers=headers_api, data=payload)

if response_api.status_code == 200:
    try:
        data = response_api.json()
        teams_data = data.get("teams", [])
        
        for team in teams_data:
            name = team.get("name")
            if name:
                registered_teams.append(name.strip())
    except ValueError:
        print("Chyba: Odpověď serveru není platný JSON.")
        sys.exit(1)
else:
    print(f"Chyba při komunikaci s API. Kód: {response_api.status_code}")
    sys.exit(1)


# --- ČÁST C: Porovnání obou seznamů ---
print("\n--- VÝSLEDEK KONTROLY ---")

missing_teams = []
registered_teams_lower = [t.lower() for t in registered_teams]

for team in upcoming_teams:
    if team.lower() not in registered_teams_lower:
        missing_teams.append(team)

if not missing_teams:
    print("✅ Vše v pořádku! Všechny týmy z webové rezervace jsou zadané v hodnotícím systému.")
else:
    print(f"Nalezeno {len(missing_teams)} chybějících týmů. Ověřuji v globální databázi...\n")
    
    # --- ČÁST D: Ověření chybějících týmů v databázi Hospodského kvízu ---
    for team in missing_teams:
        # Převedeme název týmu do formátu bezpečného pro URL (např. mezery na %20)
        search_url = f"https://www.hospodskykviz.cz/tymy/seznam?search={urllib.parse.quote(team)}"
        res_search = requests.get(search_url, headers=headers_web)
        
        found_in_db = False
        
        if res_search.status_code == 200:
            soup_search = BeautifulSoup(res_search.text, 'html.parser')
            search_tables = soup_search.find_all('table')
            
            for table in search_tables:
                first_row = table.find('tr')
                # Zkontrolujeme, zda jde o správnou tabulku s výsledky
                if first_row and 'Název domovské hospody' in first_row.text:
                    rows = table.find_all('tr')[1:] # Přeskočíme hlavičku
                    
                    if rows:
                        best_match_cols = None
                        
                        # 1. Nejprve zkusíme najít přesnou shodu jména (bez ohledu na velikost písmen)
                        for row in rows:
                            cols = row.find_all('td')
                            if len(cols) >= 2:
                                db_team_name = cols[0].text.strip()
                                if team.lower() == db_team_name.lower():
                                    best_match_cols = cols
                                    break
                        
                        # 2. Pokud se nenašla přesná shoda, vezmeme prostě první výsledek vyhledávání
                        if not best_match_cols:
                            cols = rows[0].find_all('td')
                            if len(cols) >= 2:
                                best_match_cols = cols
                        
                        # Pokud máme data, vypíšeme je
                        if best_match_cols:
                            db_team_name = best_match_cols[0].text.strip()
                            home_pub = best_match_cols[1].text.strip()
                            print(f" ⚠️  Tým '{team}' nalezen. Domovská hospoda: {home_pub} (v DB jako '{db_team_name}')")
                            found_in_db = True
                            break
        
        if not found_in_db:
            print(f" ❌ Tým '{team}' NENÍ v databázi. Tým se musí založit!")

print("\nHotovo.")

File: kontrola_tymu.scala
//> using scala "3.8.2"
//> using platform "native"
//> using dep "com.softwaremill.sttp.client4::core::4.0.19"
//> using dep "io.circe::circe-core::0.14.15"
//> using dep "io.circe::circe-parser::0.14.15"

import sttp.client4.*
import sttp.client4.curl.CurlBackend
import io.circe.parser.*
import io.circe.Json
import scala.io.StdIn

// Pomocné čisté funkce na nejvyšší úrovni souboru
def stripTags(html: String): String = html.replaceAll("(?is)<[^>]+>", "").trim

def extractTableRows(html: String): List[String] =
  "(?is)<tr[^>]*>(.*?)</tr>".r.findAllMatchIn(html).map(_.group(1)).toList

def extractCols(row: String): List[String] =
  "(?is)<t[dh][^>]*>(.*?)</t[dh]>".r.findAllMatchIn(row).map(m => stripTags(m.group(1))).toList

@main def kontrolaTymu(): Unit =
  println("Hospodský kvíz")
  print("Zadej PIN pro aktuální kvíz a stiskni Enter: ")
  scala.Console.flush()
  
  val inputPin = StdIn.readLine()
  if inputPin == null || inputPin.trim.isEmpty then
    println("Chyba: Nebyl zadán žádný PIN.")
    sys.exit(1)
    
  val pin = inputPin.trim

  val backend = CurlBackend()
  val userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

  // --- ČÁST A: Získání týmů z webu ---
  println("Stahuji seznam týmů z webu (rezervace)...")
  val urlWeb = uri"https://www.hospodskykviz.cz/hospody/u-posty-vinohrady-st/rezervace"
  val responseWeb = basicRequest.header("User-Agent", userAgent).get(urlWeb).send(backend)

  // Deklarativní zpracování HTML bez závorek
  val upcomingTeams: List[String] = responseWeb.body.toOption.toList.flatMap: html =>
    val tables = "(?is)<table[^>]*>(.*?)</table>".r.findAllMatchIn(html).map(_.group(1)).toList
    
    tables.find(t => extractTableRows(t).headOption.exists(_.contains("Název týmu")))
      .toList
      .flatMap(extractTableRows)
      .drop(1)
      .map(extractCols)
      .filter(_.length >= 2)
      .map(_.head)
      .filter(name => name != "Toto místo je zatím volné." && name.nonEmpty)

  if upcomingTeams.isEmpty then
    println("Chyba: Nepodařilo se stáhnout žádné týmy z webu nebo je stránka prázdná.")
    sys.exit(1)

  // --- ČÁST B: Získání týmů z API ---
  println(s"Stahuji seznam týmů z API pro PIN: $pin (simuluji mobilní Chrome)...")
  val urlApi = uri"https://www.hospodskykviz.cz/api/loadquiz"
  val responseApi = basicRequest
    .header("User-Agent", userAgent)
    .header("Origin", "https://hodnoceni.hospodskykviz.cz")
    .header("Accept", "application/json, text/javascript, */*; q=0.01")
    .body(Map("pin" -> pin))
    .post(urlApi)
    .send(backend)

  // Funkcionální zpracování chyb přes for-comprehension bez složených závorek
  val registeredTeamsEither: Either[String, List[String]] =
    for
      body <- responseApi.body
      json <- parse(body).left.map(_ => "Odpověď serveru není platný JSON.")
      teamsJson <- json.hcursor.downField("teams").as[List[Json]].left.map(_ => "Chybí pole 'teams'.")
    yield teamsJson.flatMap(_.hcursor.downField("name").as[String].toOption).map(_.trim)

  val registeredTeams = registeredTeamsEither match
    case Right(teams) => teams
    case Left(errorMsg) =>
      println(s"Chyba při komunikaci s API: $errorMsg")
      sys.exit(1)
      List.empty

  // --- ČÁST C: Porovnání obou seznamů ---
  println("\n--- VÝSLEDEK KONTROLY ---")
  val registeredTeamsLower = registeredTeams.map(_.toLowerCase)
  val missingTeams = upcomingTeams.filterNot(t => registeredTeamsLower.contains(t.toLowerCase))

  if missingTeams.isEmpty then
    println("✅ Vše v pořádku! Všechny týmy z webové rezervace jsou zadané v hodnotícím systému.")
  else
    println(s"Nalezeno ${missingTeams.length} chybějících týmů. Ověřuji v globální databázi...\n")

    // --- ČÁST D: Ověření chybějících týmů v DB ---
    missingTeams.foreach: team =>
      val searchUrl = uri"https://www.hospodskykviz.cz/tymy/seznam?search=$team"
      val resSearch = basicRequest.header("User-Agent", userAgent).get(searchUrl).send(backend)

      val dbMatch: Option[(String, String)] = resSearch.body.toOption.flatMap: htmlSearch =>
        val searchTables = "(?is)<table[^>]*>(.*?)</table>".r.findAllMatchIn(htmlSearch).map(_.group(1)).toList
        
        searchTables.find(t => extractTableRows(t).headOption.exists(_.contains("Název domovské hospody")))
          .map(t => extractTableRows(t).drop(1))
          .filter(_.nonEmpty)
          .flatMap: dataRows =>
            val parsedRows = dataRows.map(extractCols).filter(_.length >= 2)
            parsedRows.find(_.head.toLowerCase == team.toLowerCase)
              .orElse(parsedRows.headOption)
      .map(cols => (cols(0), cols(1)))

      dbMatch match
        case Some((dbTeamName, homePub)) =>
          println(s" ⚠️  Tým '$team' nalezen. Domovská hospoda: $homePub (v DB jako '$dbTeamName')")
        case None =>
          println(s" ❌ Tým '$team' NENÍ v databázi. Tým se musí založit!")

  println("\nHotovo.")
  println("Stiskni Enter pro ukončení...")
  StdIn.readLine()